File: Data Structures in JAVA/MaxHeap.java
package DataStructure;

import java.util.Scanner;
import java.util.*;
public class MaxHeap {
    private int CAPACITY;//常数，表示heap的大小
    private int size =0;
    private int[] array;

    public MaxHeap(int capacity){
        this.CAPACITY = capacity;
        array = new int[this.CAPACITY];
    }
    private int getLeftChildIndex(int parentIndex){
        /*获得左子树节点*/
        return 2*parentIndex+1;//0节点左子节点是1，右子节点是2,1左子节点是3，右子节点是4；
    }
    private int getRightChildIndex(int parentIndex){
        /*获得右子树节点*/
        return 2*parentIndex+2;//0节点左子节点是1，右子节点是2,1左子节点是3，右子节点是4；
    }
    private int getParentIndex(int childIndex){
        /*获得父子树节点*/
        return(childIndex-1)/2;//因为是integer，所以1/2会直接为0，3/2会直接为1；
    }

    private boolean hasLeftChild(int index){
        return getLeftChildIndex(index)<size;
    }
    private boolean hasRightChild(int index){
        return getRightChildIndex(index)<size;
    }
    private boolean hasParent(int index){
        return getParentIndex(index)>=0;
    }
    private int leftChild(int parentIndex){
        return array[getLeftChildIndex(parentIndex)];
    }
    private int rightChild(int parentIndex){
        return array[getRightChildIndex(parentIndex)];
    }
    private int parent(int child){
        return array[getParentIndex(child)];
    }

    public void insert(int item){
        if(size==CAPACITY){
            array=Arrays.copyOf(array,CAPACITY*2);
        }
        array[size]=item;
        size++;
        heapifyUp();//往上对比替换
    }
    private void heapifyUp(){
        int index = size-1;
        while(hasParent(index)&&parent(index)<array[index]){
            swap(getParentIndex(index),index);
            index = getParentIndex(index);
        }
    }

    private void swap(int parentIndex,int childIndex){
        int temp = array[parentIndex];
        array[parentIndex] = array[childIndex];
        array[childIndex] = temp;
    }

    private int poll(){
        if(size==0){
            System.out.println("the heap is empty!!!");
            return Integer.MIN_VALUE;
        }
        int element = array[0];
        array[0]=array[size-1];
        size--;
        heapifyDown();
        return element;
    }

    private void heapifyDown(){
        int index=0;
        while(hasLeftChild(index)){
            int largerChildIndex = getLeftChildIndex(index);
            if(hasRightChild(index)&&rightChild(index)>leftChild(index)){
                largerChildIndex = getRightChildIndex(index);
            }
            if(array[index]<array[largerChildIndex]){
                swap(index,largerChildIndex);
            }else{
                break;
            }
            index = largerChildIndex;
        }
    }

    private int peek(){
        if(size==0){
            throw new NoSuchElementException();//没有就抛出异常
        }
        return array[0];
    }

    public static void main(String[] args){
        /*测试代码ACM模式*/
        Scanner obj = new Scanner(System.in);
        System.out.println("Please enter the list of numbers for input, " +
                "and the last number is how many points you want to pop.\n " +
                "As ending, you need to input any characters that are not numbers.");
        List<Integer> input = new ArrayList<>();
        while(obj.hasNextInt()){
            input.add(obj.nextInt());
        }
        int poppedNum = input.remove(input.size()-1);
        MaxHeap mHeap = new MaxHeap(5);
        for(int i=0;i<input.size();i++){
            mHeap.insert(input.get(i));
        }
        for(int i=0;i<poppedNum;i++){
            System.out.println("the popped number is: "+ mHeap.poll());
        }
    }
}

File: Data Structures in Python/MaxHeap.py
class MaxHeap:
    """
    因为堆的特殊性，这里可以直接用数组来实现，而非链表
    """
    def __init__(self):
        #self.capacity = capacity # 容量，存节点数量，但因为python的list是无限容量的，所以不需要capacity
        self.size = 0 # 当前节点数量
        self.array = list() # 存当前节点的数组


    """
    接下来是获得各个节点的index
    """
    def getLeftChildIndex(self,parentIndex):
        """
        通过计算获得孩子的节点
        """
        return 2*parentIndex +1

    def getRightChildIndex(self,parentIndex):
        return 2*parentIndex.key +2

    def getParentIndex(self,childIndex):
        return (childIndex-1)//2#底板除

    def hasLeftChild(self,index:int):
        return self.getLeftChildIndex(index)<self.size

    def hasRightChild(self,index):
        return self.getLeftChildIndex(index)<self.size

    def hasParent(self,index):
        return (index - 1) // 2 >=0 # 底板除

    """
    接下来是取得节点的值
    """
    def getLeftChildValue(self, parentIndex):
        return self.array[self.getLeftChildIndex(parentIndex)]

    def getRightChildValue(self, parentIndex):
        return self.array[self.getRightChildIndex(parentIndex)]

    def getParentValue(self,childIndex):
        return self.array[self.getParentIndex(childIndex)]

    """
    接下来是各种直接对heap的操作
    """

    def insert(self,item):
        self.array.append(item)
        self.size+=1
        self.heapifyUp()


    def heapifyUp(self):
        index = self.size - 1 #因为在insert里size永远在插入之后加1，所以size比index大1
        while self.hasParent(index) and self.getParentValue(index) < self.array[index]:
            self.swap(self.getParentIndex(index),index)#互换
            index = self.getParentIndex(index)#更新子节点

    def swap(self,parentIndex,childIndex):
        temp = self.array[parentIndex]
        self.array[parentIndex] = self.array[childIndex]
        self.array[childIndex] = temp

    def poll(self):
        """
        移除根节点，把最大数移到根节点处
        """
        if self.size == 0:
            return
        element = self.array[0]
        self.array[0]=self.array[self.size-1]#将最后输入的数字放到根节点
        self.size -=1
        self.heapifyDown()


    def heapifyDown(self):
        index =0
        while self.hasLeftChild(index):
            largerChildIndex = self.getLeftChildIndex(index)
            if self.hasRightChild(index) and self.getRightChildValue(index) > self.getLeftChildValue(index):
                largerChildIndex = self.getRightChildValue(index)

            if self.array[index] < self.array[largerChildIndex]:
                self.swap(index, largerChildIndex)
            else:
                break

            index = largerChildIndex

    def peek(self):
        if self.size ==0:
            return
        else:
            return self.array[0]


